docs/hazmat/primitives/asymmetric/dsa.rst

```diff
@@ -74,8 +74,55 @@ provider.
     ...     key_size=1024,
     ...     backend=default_backend()
     ... )
-    >>> signer = private_key.signer(hashes.SHA256())
     >>> data = b"this is some data I'd like to sign"
+
+Signing using a :class:`cryptography.hazmat.primitives.hashes.HashContext` (preferred method):
+
+.. doctest::
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> signature = private_key.sign(digester)
+
+One-line signing:
+
+.. doctest::
+
+    >>> signature = private_key.sign(data, hashes.SHA256())
+
+Signing a precomputed digest:
+
+.. warning::
+
+    Only use this for digests, do NOT sign raw data. The data will get truncated
+    to the size of DSA's "q" parameter, which is just a few bytes.
+
+.. doctest::
+
+    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
+    >>> digester.update(data)
+    >>> digest = digester.finalize()
+    >>> signature = private_key.sign(digest, already_hashed=True)
+
+Specifying the per-message key k:
```

Comment:
There's no reason for it. pyca/cryptography, we shouldn't be giving people
more tools to shoot themselves in the foot than is _absolutely_ necessary.

On Wed, Jan 6, 2016 at 1:33 PM, joernheissler <notifications@github.com>
wrote:

> In docs/hazmat/primitives/asymmetric/dsa.rst
> <https://github.com/pyca/cryptography/pull/2642#discussion_r48991323>:
>
> > +
> > +Signing a precomputed digest:
> > +
> > +.. warning::
> > +
> > +    Only use this for digests, do NOT sign raw data. The data will get truncated
> > +    to the size of DSA's "q" parameter, which is just a few bytes.
> > +
> > +.. doctest::
> > +
> > +    >>> digester = hashes.Hash(hashes.SHA256(), backend=default_backend())
> > +    >>> digester.update(data)
> > +    >>> digest = digester.finalize()
> > +    >>> signature = private_key.sign(digest, already_hashed=True)
> > +
> > +Specifying the per-message key k:
>
> For debugging and educational purposes. And perhaps to allow users to
> implement stuff like rfc6979. Openssl implemented something similar which
> uses the private key + message digest do seed a PRNG. But it's
> non-deterministic. I'm not saying that this approach is bad, but it's not
> rfc6979.
>
> What's wrong with supporting a dangerous API? For one, it's all in
> "hazmat". And with the big fat warning even a stupid user shouldn't
> accidentally use this function.
> Perhaps change "warning" to "dangerous"?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/pyca/cryptography/pull/2642/files#r48991323>.
>



-- 
"I disapprove of what you say, but I will defend to the death your right to
say it." -- Evelyn Beatrice Hall (summarizing Voltaire)
"The people's good is the highest law." -- Cicero
GPG Key fingerprint: 125F 5C67 DFE9 4084